### PR TITLE
Cache subpopulations

### PR DESCRIPTION
**Pull request by alxdarksage**
Some logic that is more business logic than persistence was moved into the service, which makes it far easier to cache at the level of persistence only.

### CODE REVIEW

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java [hunk 1]
```diff
@@ -220,6 +221,21 @@ public void removeStudy(String identifier) {
         return null;
     }
     
+    public <T> T getObject(String cacheKey, TypeReference<T> typeRef) {
```
Comment (zaro0508) [inline app/org/sagebionetworks/bridge/cache/CacheProvider.java:224]: This looks ok, but it doesn't seem to be used anywhere?

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationDaoMockTest.java [hunk 2]
```diff
@@ -125,43 +124,44 @@ public void getSubpopulationConstructsCriteriaIfNotSaved() {
...
+        doReturn(defaultSubpop).when(dao).getSubpopulation(TEST_STUDY, SUBPOP_GUID);
+        doReturn(Criteria.create()).when(criteriaDao).getCriteria(any());
+        
+        dao.deleteSubpopulation(TEST_STUDY, SUBPOP_GUID, true, false);
     }
     
     @Test
-    public void logicalDeleteSubpopulationDoesNotDeleteCriteria() {
-        dao.deleteSubpopulation(TEST_STUDY, SUBPOP_GUID, false);
+    public void allowDeleteOfDefault() {
```
Comment (zaro0508) [inline test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationDaoMockTest.java]: Looks like this checks physical delete of default subpopulation.  Is there a case where you would do a logical delete of default?  it doesn't look like that case is covered.

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/cache/CacheProvider.java:224]: It is used in SubpopulationService line 172 to get the cached list of subpopulations

File: test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationDaoMockTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/dynamodb/DynamoSubpopulationDaoMockTest.java]: Added test that logical delete also cannot delete the default (unless you're calling from the delete everything service method which sets the flag to override this, used for tests).

File: app/org/sagebionetworks/bridge/cache/CacheProvider.java (hunk 1 above)
Comment (zaro0508) [reply app/org/sagebionetworks/bridge/cache/CacheProvider.java:224]: thanks.  i really should figure out how to setup intelliji to make it correctly compile the project so i can use the 'usage' feature.

**Review by zaro0508 — APPROVED** [review]
LGTM

**Pull request merged**